crates/bifrost/src/providers/replicated_loglet/tasks/digests.rs

```diff
@@ -0,0 +1,358 @@
+        if entry.is_trim_gap() {
+            debug!(
+                loglet_id = %self.loglet_id,
+                "Observed a trim gap from offset {} to {} while repairing the tail of the loglet, \
+                 those offsets will be considered repaired",
+                offset,
+                entry.trim_gap_to_sequence_number().unwrap(),
+            );
+            self.update_from_offset(entry.next_sequence_number());
+            return Ok(());
+        }
+
+        let known_copies = self
+            .offsets_under_repair
+            .get(&offset)
+            .expect("repairing an offset that we have not truncated");
+
+        // See how many copies do we need to do to achieve write quorum.
+        let fixup_nodes = self
+            .spread_selector
+            .select_fixups(
+                known_copies,
+                &mut thread_rng(),
+                &networking.metadata().nodes_config_ref(),
+                &NodeSet::empty(),
+            )
+            // what do we do if we can't generate a spread? nodes are in data-loss or readonly, or
+            // whatever state.
+            // We definitely cannot proceed but should we retry? let's only do that when we need to,
+            // for now, we bail.
+            .map_err(OperationError::retryable)?;
+
+        trace!(
+            loglet_id = %self.loglet_id,
+            %offset,
+            "Repairing record, existing copies on {} and fixup nodes are {}",
+            known_copies,
+            fixup_nodes,
+        );
+
+        // We handled trim gaps before this point.
+        let record = entry.into_record().expect("must be a data record");
+
+        let mut replication_checker = NodeSetChecker::new(
+            self.spread_selector.nodeset(),
+            &networking.metadata().nodes_config_ref(),
+            self.spread_selector.replication_property(),
+        );
+        // record is already replicated on those nodes
+        replication_checker.set_attribute_on_each(known_copies, || true);
+
+        let payloads = vec![record].into();
+
+        let msg = Store {
+            // As we send store messages, we consider the from_offset a reliable source of
+            // global_known_loglet. This allows log-servers to accept those writes if they were still
+            // behind.
+            header: LogServerRequestHeader::new(self.loglet_id, self.start_offset),
+            timeout_at: None,
+            // Must be set to bypass the seal
+            flags: StoreFlags::IgnoreSeal,
+            first_offset: offset,
+            sequencer,
+            known_archived: LogletOffset::INVALID,
+            payloads,
+        };
+        // We run stores as tasks because we'll wait only for the necessary write-quorum but the
+        // rest of the stores can continue in the background as best-effort replication (if the
+        // spread selector strategy picked extra nodes)
```

Comment:
Unrelated: Can it become a problem if we accumulate network send tasks that are awaiting a response which won't come because the other node has died? I don't expect this to happen often but over time it could result into a memory leak.

Reply:
The task will give up once it exhausts our rpc retry policy (which is finite by default) but the risk exists if someone changed this to infinite retries.